<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><title>R: Estimation and correction of GC bias in short-read DNA
sequencing</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link rel="stylesheet" type="text/css" href="R.css">
</head><body>
<h1> Estimation and correction of GC bias in short-read DNA
sequencing
<img class="toplogo" src="../../../doc/html/logo.jpg" alt="[R logo]">
</h1>
<hr>
<div align="center">
<a href="../../../doc/html/packages.html"><img src="../../../doc/html/left.jpg" alt="[Up]" width="30" height="30" border="0"></a>
<a href="../../../doc/html/index.html"><img src="../../../doc/html/up.jpg" alt="[Top]" width="30" height="30" border="0"></a>
</div><h2>Documentation for package &lsquo;GCcorrect&rsquo; version 0.96</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
<li><a href="../demo">Code demos</a>.  Use <a href="../../utils/help/demo">demo()</a> to run them.</li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td width="25%"><a href="GCcorrect-package.html">GCcorrect-package</a></td>
<td>Estimation and correction of GC bias in short-read DNA sequencing</td></tr>
<tr><td width="25%"><a href="basicPlots.html">basicPlots</a></td>
<td>Plots of reads</td></tr>
<tr><td width="25%"><a href="binReads.html">binReads</a></td>
<td>Count reads in each bin.</td></tr>
<tr><td width="25%"><a href="compDataRefPlots.html">compDataRefPlots</a></td>
<td>Exploratory Plots Comparing Reads and Chromosome</td></tr>
<tr><td width="25%"><a href="condMeanAndPredict.html">condMeanAndPredict</a></td>
<td>Estimates the GC bias, and create a prediction vector</td></tr>
<tr><td width="25%"><a href="findLongZeros.html">findLongZeros</a></td>
<td>Find Long Zeros</td></tr>
<tr><td width="25%"><a href="GCcorrect-package.html">GCcorrect</a></td>
<td>Estimation and correction of GC bias in short-read DNA sequencing</td></tr>
<tr><td width="25%"><a href="getCondMean.html">getCondMean</a></td>
<td>Conditional Mean</td></tr>
<tr><td width="25%"><a href="makeGCLens.html">makeGCLens</a></td>
<td>Simultanously estimate GC curves for multiple GC windows</td></tr>
<tr><td width="25%"><a href="meanLine.html">meanLine</a></td>
<td>Return Mean for each Bin</td></tr>
<tr><td width="25%"><a href="getCondMean.html">plotCondMean</a></td>
<td>Conditional Mean</td></tr>
<tr><td width="25%"><a href="makeGCLens.html">plotGCLens</a></td>
<td>Simultanously estimate GC curves for multiple GC windows</td></tr>
<tr><td width="25%"><a href="prepareChrom.html">prepareChrom</a></td>
<td>Prepare Chromosome Information Structure</td></tr>
<tr><td width="25%"><a href="prepareGCChrom.html">prepareGCChrom</a></td>
<td>Calculate Line of GCs of moving window</td></tr>
<tr><td width="25%"><a href="prepareReads.html">prepareReads</a></td>
<td>Prepare Read Structure</td></tr>
<tr><td width="25%"><a href="sampleChrom.html">sampleChrom</a></td>
<td>Create sample of chromosome positions for estimating GC effect.</td></tr>
<tr><td width="25%"><a href="makeGCLens.html">scoreGCLens</a></td>
<td>Simultanously estimate GC curves for multiple GC windows</td></tr>
<tr><td width="25%"><a href="unEvenBin.html">unEvenBin</a></td>
<td>Calculate Sum for Uneven Bins</td></tr>
</table>
</body></html>
