<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><title>R: Transcript mapping with high-density oligonucleotide tiling
arrays</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link rel="stylesheet" type="text/css" href="R.css">
</head><body>
<h1> Transcript mapping with high-density oligonucleotide tiling
arrays
<img class="toplogo" src="../../../doc/html/logo.jpg" alt="[R logo]">
</h1>
<hr>
<div align="center">
<a href="../../../doc/html/packages.html"><img src="../../../doc/html/left.jpg" alt="[Up]" width="30" height="30" border="0"></a>
<a href="../../../doc/html/index.html"><img src="../../../doc/html/up.jpg" alt="[Top]" width="30" height="30" border="0"></a>
</div><h2>Documentation for package &lsquo;tilingArray&rsquo; version 1.32.0</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
<li><a href="../doc/index.html">Overview of user guides and package vignettes</a>; browse <a href="../doc">directory</a>.</li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td width="25%"><a href="tilingArray-package.html">tilingArray-package</a></td>
<td>tilingArray package overview</td></tr>
<tr><td width="25%"><a href="PMindex.html">BGindex</a></td>
<td>Find the index of the exact match (PM) or background probes from a probeAnno environment</td></tr>
<tr><td width="25%"><a href="breakpointsPretend.html">breakpoints.breakpointsPretend</a></td>
<td>Accessor methods for breakpointsPretend objects - not to be called by the user.</td></tr>
<tr><td width="25%"><a href="comparisonPlot.html">comparisonPlot</a></td>
<td>Plot a vertical layout of panels for the comparison of different along-chromosome profiles.</td></tr>
<tr><td width="25%"><a href="class-segmentation.html">confint</a></td>
<td>The class segmentation represents a segmentation result.</td></tr>
<tr><td width="25%"><a href="class-segmentation.html">confint-method</a></td>
<td>The class segmentation represents a segmentation result.</td></tr>
<tr><td width="25%"><a href="costMatrix.html">costMatrix</a></td>
<td>Segmentation cost matrix</td></tr>
<tr><td width="25%"><a href="findsegments.html">findSegments</a></td>
<td>Fit a piecewise constant curve to a sequence of numbers - OBSOLETE, please use function segment instead.</td></tr>
<tr><td width="25%"><a href="gffSub.html">gffSub</a></td>
<td>Example of a genomic feature object</td></tr>
<tr><td width="25%"><a href="class-segmentation.html">logLik</a></td>
<td>The class segmentation represents a segmentation result.</td></tr>
<tr><td width="25%"><a href="class-segmentation.html">logLik-method</a></td>
<td>The class segmentation represents a segmentation result.</td></tr>
<tr><td width="25%"><a href="normalizeByReference.html">normalizeByReference</a></td>
<td>Probe-specific normalization of hybridization intensities from an oligonucleotide microarray</td></tr>
<tr><td width="25%"><a href="otherStrand.html">otherStrand</a></td>
<td>Return the name of the opposite strand</td></tr>
<tr><td width="25%"><a href="class-segmentation.html">plot-method</a></td>
<td>The class segmentation represents a segmentation result.</td></tr>
<tr><td width="25%"><a href="plotAlongChrom.html">plotAlongChrom</a></td>
<td>Plot signals and segmentation for a region of a chromosome</td></tr>
<tr><td width="25%"><a href="plotAlongChromLegend.html">plotAlongChromLegend</a></td>
<td>Plot a legend for genomic features</td></tr>
<tr><td width="25%"><a href="plotFeatures.html">plotFeatures</a></td>
<td>Plot genomic features for a region along a chromosome</td></tr>
<tr><td width="25%"><a href="plotpenLL.html">plotPenLL</a></td>
<td>Plot the log-likelihood and penalized log-likelihoods (AIC, BIC)</td></tr>
<tr><td width="25%"><a href="plotSegmentationDots.html">plotSegmentationDots</a></td>
<td>Plot points for a region along a chromosome</td></tr>
<tr><td width="25%"><a href="plotSegmentationHeatmap.html">plotSegmentationHeatmap</a></td>
<td>Plot a heatmap diagram for a region along a chromosome</td></tr>
<tr><td width="25%"><a href="PMindex.html">PMindex</a></td>
<td>Find the index of the exact match (PM) or background probes from a probeAnno environment</td></tr>
<tr><td width="25%"><a href="posMin.html">posMin</a></td>
<td>Find the smallest positive number in a vector</td></tr>
<tr><td width="25%"><a href="qcPlots.html">qcPlots</a></td>
<td>Generate simple diagnostic plots for Affymetrix tiling array data</td></tr>
<tr><td width="25%"><a href="readCel2eSet.html">readCel2eSet</a></td>
<td>Read celfiles into an ExpressionSet object.</td></tr>
<tr><td width="25%"><a href="breakpointsPretend.html">residuals.breakpointsPretend</a></td>
<td>Accessor methods for breakpointsPretend objects - not to be called by the user.</td></tr>
<tr><td width="25%"><a href="sampleStep.html">sampleStep</a></td>
<td>Sampling of ascending numbers to ensure minimal spacing.</td></tr>
<tr><td width="25%"><a href="segChrom.html">segChrom</a></td>
<td>Fit a piecewise constant curve to along chromosome data (wrapper function)</td></tr>
<tr><td width="25%"><a href="segment.html">segment</a></td>
<td>Fit a piecewise constant curve: segmentation by dynamic programming</td></tr>
<tr><td width="25%"><a href="class-segmentation.html">segmentation-class</a></td>
<td>The class segmentation represents a segmentation result.</td></tr>
<tr><td width="25%"><a href="segnf.html">segnf</a></td>
<td>Example of a segmentation output object</td></tr>
<tr><td width="25%"><a href="class-segmentation.html">show-method</a></td>
<td>The class segmentation represents a segmentation result.</td></tr>
<tr><td width="25%"><a href="tilingArray-package.html">tilingArray</a></td>
<td>tilingArray package overview</td></tr>
</table>
</body></html>
